spring-boot-starter/spring-boot-starter-discovery/src/main/java/io/servicecomb/springboot/starter/discovery/CseDiscoveryClient.java

```diff
@@ -45,21 +44,18 @@ public String description() {
 
   @Override
   public List<ServiceInstance> getInstances(final String serviceId) {
-    List<ServiceInstance> instances = new ArrayList<>();
-    ServiceRegistryClient client = RegistryUtils.getServiceRegistryClient();
-    String appId = RegistryUtils.getAppId();
-    ReferenceConfig referenceConfig = consumerProviderManager.getReferenceConfig(serviceId);
-    String versionRule = referenceConfig.getMicroserviceVersionRule();
-    String cseServiceID = client.getMicroserviceId(appId, serviceId, versionRule);
-    List<MicroserviceInstance> cseServices = client.getMicroserviceInstance(cseServiceID, cseServiceID);
-    if (null != cseServices && !cseServices.isEmpty()) {
-      for (MicroserviceInstance instance : cseServices) {
-        List<String> eps = instance.getEndpoints();
-        for (String ep : eps) {
-          URIEndpointObject uri = new URIEndpointObject(ep);
-          instances.add(new DefaultServiceInstance(instance.getServiceId(), uri.getHostOrIp(),
-              uri.getPort(), false));
-        }
+    DiscoveryContext context = new DiscoveryContext();
+    context.setInputParameters(serviceId);
+    VersionedCache serversVersionedCache = discoveryTree.discovery(context,
+        RegistryUtils.getAppId(),
+        serviceId,
+        DefinitionConst.VERSION_RULE_ALL);
+    Map<String, MicroserviceInstance> servers = serversVersionedCache.data();
```

Comment:
convert MicroserviceInstance results to List<ServcieInstance> can be a filter of discoveryTree
this can avoid build List<ServiceInstance> for every invocation.

Reply:
create a issue to track this : SCB-207